File: implementacao java/src/Dominio/Partido.java
package Dominio;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;


public class Partido implements Comparable<Partido>{
    private int numeroPartido;
    private String siglaPartido;
    private int votosNominais;
    private int votosLegenda;
    private int votosTotais;
    private Set<Candidato> candidatosEleitos = new HashSet<>();
    private Set<Candidato> candidatos = new HashSet<>();

    public Partido(int numeroPartido, String siglaPartido){
        this.numeroPartido = numeroPartido;
        this.siglaPartido = siglaPartido;
    }

    public void atualizaAtributos(int nominais, int legenda){

        this.votosNominais += nominais;
        this.votosLegenda += legenda;
        
        this.votosTotais = this.votosNominais + this.votosLegenda;
    }

    public void adicionarCandidatoEleito(Candidato candidato){
        this.candidatosEleitos.add(candidato);

    }

    public void adicionarCandidato(Candidato candidato){
        this.candidatos.add(candidato);

    }

    public int candidatoMaisVotado(){

        if (this.candidatos.isEmpty()) {
            return -1; 
        }

        List<Candidato> listaOrdenavel = new ArrayList<>(this.candidatos);
        Collections.sort(listaOrdenavel);    

        return listaOrdenavel.get(0).getQtdVotosRecebidos();

    }


    public Set<Candidato> getCandidatos() {
        return new HashSet<>(this.candidatos);
    }


    public Set<Candidato> getCandidatosEleitos(){
        return new HashSet<>(this.candidatosEleitos);
    } 
    

    public String getSiglaPartido() {
        return siglaPartido;
    }

    public int getVotosTotais() {
        return this.votosTotais;
    }

    public int getNumeroPartido() {
        return numeroPartido;
    }

    public int getVotosNominais() {
        return votosNominais;
    }

    public int getVotosLegenda(){
        return votosLegenda;
    }

    @Override
    public int compareTo(Partido partido){
        int compCriterioI = Integer.compare(partido.getVotosTotais(), this.votosTotais);

        if(compCriterioI == 0){
            compCriterioI = Integer.compare(this.getNumeroPartido(), partido.getNumeroPartido());
        }

        return compCriterioI;

    }

     
}

File: implementacao c++/Partido.cpp
#include "Partido.hpp"
#include "Candidato.hpp"
#include <vector>
#include <algorithm> // Para std::max_element



Partido::Partido(int numeroPartido, const std::string &siglaPartido){

    this->numeroPartido = numeroPartido;
    this->siglaPartido = siglaPartido;
    
    this->votosLegenda = 0;
    this->votosNominais = 0;
    this->votosTotais = 0;
}

void Partido::atualizaAtributos(int nominais, int legenda){
    this->votosNominais += nominais;
    this->votosLegenda += legenda;
    
    this->votosTotais = this->votosNominais + this->votosLegenda;
}

std::set<const Candidato*> Partido::getCandidatosEleitos(){
    return this->candidatosEleitos;
}

int Partido::getNumeroPartido() const{

    return this->numeroPartido;

}

std::string Partido::getSiglaPartido() const{

    return this->siglaPartido;

}

int Partido::getVotosTotais() const{
    return this->votosTotais;
}

int Partido::getVotosNominais()const{
    return this->votosNominais;
}

int Partido::getVotosLegenda()const{
    return this->votosLegenda;
}

bool Partido::operator<(Partido const &outro)const{ 
    if(this->votosTotais != outro.votosTotais){
        return this->votosTotais > outro.votosTotais;
    }

    return this->numeroPartido < outro.numeroPartido;
}


void Partido::adicionarCandidato(const Candidato *candidato){

    this->candidatos.insert(candidato);

}

void Partido::adicionarCandidatoEleito(const Candidato *candidato){

    this->candidatosEleitos.insert(candidato);

}

const Candidato* Partido::candidatoMaisVotado()const {
   
    if (this->candidatos.empty()) {
        return nullptr;
    }

    auto comparador = [](const Candidato* a, const Candidato* b) {
        if(a->getVotos() != b->getVotos()){
            return a->getVotos() < b->getVotos();
        }
        return a->getIdade() < b->getIdade();
    };

    auto it_mais_votado = std::max_element(this->candidatos.begin(), this->candidatos.end(), comparador);

    return *it_mais_votado;
}


const Candidato* Partido::candidatoMenosVotado() const {
   
    if (this->candidatos.empty()) {
        return nullptr;
    }

    auto comparador = [](const Candidato* a, const Candidato* b) {
        if(a->getVotos() != b->getVotos()){
            return a->getVotos() < b->getVotos();
        }
        return a->getIdade() < b->getIdade();
    };

    auto it_mais_votado = std::min_element(this->candidatos.begin(), this->candidatos.end(), comparador);

    return *it_mais_votado;
}

